src/traces/icicle/defaults.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+var Lib = require('../../lib');
+var attributes = require('./attributes');
+var Color = require('../../components/color');
+var handleDomainDefaults = require('../../plots/domain').defaults;
+var handleText = require('../bar/defaults').handleText;
+var TEXTPAD = require('../bar/constants').TEXTPAD;
+
+var Colorscale = require('../../components/colorscale');
+var hasColorscale = Colorscale.hasColorscale;
+var colorscaleDefaults = Colorscale.handleDefaults;
+
+module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
+    function coerce(attr, dflt) {
+        return Lib.coerce(traceIn, traceOut, attributes, attr, dflt);
+    }
+
+    var labels = coerce('labels');
+    var parents = coerce('parents');
+
+    if(!labels || !labels.length || !parents || !parents.length) {
+        traceOut.visible = false;
+        return;
+    }
+
+    var vals = coerce('values');
+    if(vals && vals.length) {
+        coerce('branchvalues');
+    } else {
+        coerce('count');
+    }
+
+    coerce('level');
+    coerce('maxdepth');
+
+    coerce('tiling.orientation');
+    coerce('tiling.flip');
+    coerce('tiling.pad');
+
+    var text = coerce('text');
+    coerce('texttemplate');
+    if(!traceOut.texttemplate) coerce('textinfo', Array.isArray(text) ? 'text+label' : 'label');
+
+    coerce('hovertext');
+    coerce('hovertemplate');
+
+    var hasPathbar = coerce('pathbar.visible');
+
+    var textposition = 'auto';
+    handleText(traceIn, traceOut, layout, coerce, textposition, {
+        hasPathbar: hasPathbar,
+        moduleHasSelected: false,
+        moduleHasUnselected: false,
+        moduleHasConstrain: false,
+        moduleHasCliponaxis: false,
+        moduleHasTextangle: false,
+        moduleHasInsideanchor: false
+    });
+    coerce('textposition');
+
+    var lineWidth = coerce('marker.line.width');
+    if(lineWidth) coerce('marker.line.color', layout.paper_bgcolor);
+
+    coerce('marker.colors');
+    var withColorscale = traceOut._hasColorscale = (
+        hasColorscale(traceIn, 'marker', 'colors') ||
+        (traceIn.marker || {}).coloraxis // N.B. special logic to consider "values" colorscales
+    );
+    if(withColorscale) {
+        colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
+    }
+
+    if(withColorscale) {
+        colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
+    }
+
```

Comment:
@archmoj are the if statement's from `70 - 76` supposed to be repeated twice?

Reply:
No. Could you please use git blame to find which commit added them?

Reply:
@archmoj 

Git blame:
```
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  70)     if(withColorscale) {
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  71)         colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  72)     }
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  73)
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  74)     if(withColorscale) {
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  75)         colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
4f53fe1aa3 (archmoj 2021-02-22 22:40:18 -0500  76)     }
```

Commit `4f53fe1aa3`:
```
$ git log 4f53fe1aa3
commit 4f53fe1aa3166c04d308f06dd21e0af4be3d3cae
Author: archmoj <mojtaba@plot.ly>
Date:   Mon Feb 22 22:40:18 2021 -0500

    start icicle trace by duplicating and adjusting treemap code
```
Looks like it's from this commit: https://github.com/plotly/plotly.js/pull/5546/commits/b3c9c08694150261960e973b3c0eab66fa2a5142#diff-0bcbcd94d6a9283196efb8da96d88649ade581d6132132d82ec2cd518d0b82c8R75-R90

but I think you fixed it?

Reply:
We should not have this twice. So let's remove the duplicate code.

Reply:
Also please open a separate PR to remove this duplicate in `treemap`.
https://github.com/plotly/plotly.js/blob/623fcd1fea9d9bfb86e5e0d44d8047cd8636881c/src/traces/treemap/defaults.js#L88-L90

Reply:
@archmoj 

Hi, I'm looking at the `treemap` code and just had a question about deleting the `if` statement.

If I understand correctly the `coerce()` function is used to ensure that an attribute is `valid` right? It's just a check and nothing more or am I wrong? 

If I'm correct then deleting the `if(withColorscale)` won't matter, but if `coerce()` modifies the attributes in some way, then the second `if` statement may be necessary:

```
   coerce('marker.pad.t', bottomText ? headerSize / 4 : headerSize);
    coerce('marker.pad.l', headerSize / 4);
    coerce('marker.pad.r', headerSize / 4);
    coerce('marker.pad.b', bottomText ? headerSize : headerSize / 4);

    if(withColorscale) {
        colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
    }
```

Entire code block:
https://github.com/plotly/plotly.js/blob/623fcd1fea9d9bfb86e5e0d44d8047cd8636881c/src/traces/treemap/defaults.js#L75-L90
```
    if(withColorscale) {
        colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
    } else {
        coerce('marker.depthfade', !(colors || []).length);
    }

    var headerSize = traceOut.textfont.size * 2;

    coerce('marker.pad.t', bottomText ? headerSize / 4 : headerSize);
    coerce('marker.pad.l', headerSize / 4);
    coerce('marker.pad.r', headerSize / 4);
    coerce('marker.pad.b', bottomText ? headerSize : headerSize / 4);

    if(withColorscale) {
        colorscaleDefaults(traceIn, traceOut, layout, coerce, {prefix: 'marker.', cLetter: 'c'});
    }
```

Reply:
Coerce also provides default values. I think you can safely remove one of those duplicate calls.

Reply:
@archmoj made a PR for `treemap` here: https://github.com/plotly/plotly.js/pull/5565